// react-render-counter/RenderCounter/withRenderCount.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _recompose = require('recompose');

var _hocs = require('../hocs');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var componentWillUpdate = function componentWillUpdate(_ref) {
  var self = _ref.self,
      onRerender = _ref.onRerender;
  return onRerender(self.count + 1);
};

var getInitialSelf = function getInitialSelf(_ref2) {
  var initialCount = _ref2.initialCount;
  return { count: initialCount };
};

var onRerender = function onRerender(_ref3) {
  var self = _ref3.self;
  return function (count) {
    self.count = count; // eslint-disable-line no-param-reassign
  };
};

var propTypes = exports.propTypes = {
  initialCount: _propTypes2.default.number,
  onRerender: _propTypes2.default.func
};

var getOverridedDefaultProps = function getOverridedDefaultProps() {
  var _ref4 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      count = _ref4.count;

  return {
    initialCount: typeof count === 'undefined' ? 1 : count
  };
};

exports.default = function (Component) {
  return (0, _recompose.compose)((0, _hocs.omitPropTypes)('count'), (0, _hocs.extendStatics)({
    displayName: 'withRenderCount',
    propTypes: propTypes,
    defaultProps: getOverridedDefaultProps(Component.defaultProps)
  }), (0, _hocs.copyStatics)(Component), (0, _hocs.withThis)(getInitialSelf), (0, _hocs.embedHandlers)({ onRerender: onRerender }), (0, _recompose.lifecycle)({ componentWillUpdate: componentWillUpdate }), (0, _recompose.flattenProp)('self'), (0, _hocs.omitProps)(['self', 'onRerender', 'initialCount']))(Component);
};
